flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategyTest.java

```diff
@@ -0,0 +1,194 @@
+
+		for (int i = 0; i < 3; i++) {
+			schedulingTopology.addSchedulingVertex(new TestingSchedulingVertex(jobVertex1.getID(), i));
+			schedulingTopology.addSchedulingVertex(new TestingSchedulingVertex(jobVertex2.getID(), i));
+		}
+
+		TestingSchedulerOperation testingSchedulerOperation = new TestingSchedulerOperation();
+		LazyFromSourcesSchedulingStrategy schedulingStrategy = new LazyFromSourcesSchedulingStrategy(
+			testingSchedulerOperation,
+			schedulingTopology,
+			graph);
+
+		schedulingStrategy.startScheduling();
+
+		assertEquals(3, testingSchedulerOperation.getScheduledVertices().size());
+	}
+
+	/**
+	 * Tests that when on execution state change will start available downstream vertices.
+	 * vertex#0    vertex#1
+	 *       \     /
+	 *        \   /
+	 *         \ /
+	 *  (BLOCKING, ALL)
+	 *     vertex#3      vertex#2
+	 *           \        /
+	 *            \      /
+	 *             \    /
+	 *          (BLOCKING, ANY)
+	 *             vertex#4
+	 *                |
+	 *                |
+	 *                |
+	 *            (PIPELINED)
+	 *             vertex#5
+	 */
+	@Test
+	public void testOnExecutionStateChange() {
+		TestingSchedulingTopology schedulingTopology = new TestingSchedulingTopology();
+		JobVertex[] jobVertices = new JobVertex[6];
+
+		for (int i = 0; i < 6; i++) {
+			jobVertices[i] = new JobVertex("vertex#" + i);
+		}
+
+		jobVertices[3].connectNewDataSetAsInput(jobVertices[0], DistributionPattern.ALL_TO_ALL, ResultPartitionType.BLOCKING);
+		jobVertices[3].connectNewDataSetAsInput(jobVertices[1], DistributionPattern.ALL_TO_ALL, ResultPartitionType.BLOCKING);
+		jobVertices[4].connectNewDataSetAsInput(jobVertices[3], DistributionPattern.ALL_TO_ALL, ResultPartitionType.BLOCKING);
+		jobVertices[4].connectNewDataSetAsInput(jobVertices[2], DistributionPattern.ALL_TO_ALL, ResultPartitionType.BLOCKING);
+		jobVertices[5].connectNewDataSetAsInput(jobVertices[4], DistributionPattern.ALL_TO_ALL, ResultPartitionType.PIPELINED);
+
+		jobVertices[3].setInputDependencyConstraint(InputDependencyConstraint.ALL);
+		jobVertices[4].setInputDependencyConstraint(InputDependencyConstraint.ANY);
+
+		IntermediateDataSet[] dataSets = new IntermediateDataSet[6];
+		for (int i = 0; i < 5; i++) {
+			dataSets[i] = jobVertices[i].getProducedDataSets().get(0);
+		}
+
+		TestingSchedulingVertex[][] vertices = new TestingSchedulingVertex[6][2];
+		TestingSchedulingResultPartition[][] partitions = new TestingSchedulingResultPartition[6][2];
+		for (int i = 5; i >= 0; i--) {
+			for (int j = 0; j < 2; j++) {
+				vertices[i][j] = new TestingSchedulingVertex(jobVertices[i].getID(), j);
+				schedulingTopology.addSchedulingVertex(vertices[i][j]);
+
+				if (i != 5) {
+					ResultPartitionType type = i == 4 ? ResultPartitionType.PIPELINED : ResultPartitionType.BLOCKING;
+					partitions[i][j] = new TestingSchedulingResultPartition(dataSets[i].getId(), new IntermediateResultPartitionID(),
```

Comment:
I think it is not ideal that it is currently easily possible to create contradicting topologies. For example, it is possible to create an instance of `TestingSchedulingResultPartition` that links to a producer vertex with empty _"producedResultPartitions"_. This is because we currently have to create the links between partitions and vertices in two separate steps.

Reply:
OK, I will refine the structure.